src/mscorlib/src/System/Diagnostics/Eventing/EventPipe.cs

```diff
@@ -0,0 +1,155 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Security;
+using Microsoft.Win32;
+
+namespace System.Diagnostics.Tracing
+{
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct EventPipeProviderConfiguration
+    {
+        [MarshalAs(UnmanagedType.LPWStr)]
+        private string m_providerName;
+        private UInt64 m_keywords;
+
+        internal EventPipeProviderConfiguration(
+            string providerName,
+            UInt64 keywords)
+        {
+            if(string.IsNullOrEmpty(providerName))
+            {
+                throw new ArgumentNullException(nameof(providerName));
+            }
+            m_providerName = providerName;
+            m_keywords = keywords;
+        }
+
+        internal string ProviderName
+        {
+            get { return m_providerName; }
+        }
+
+        internal UInt64 Keywords
+        {
+            get { return m_keywords; }
+        }
+    }
+
+    internal sealed class EventPipeConfiguration
+    {
+        private string m_outputFile;
+        private uint m_circularBufferSizeInMB;
+        private uint m_level;
+        private List<EventPipeProviderConfiguration> m_providers;
+
+        internal EventPipeConfiguration(
+            string outputFile,
+            uint circularBufferSizeInMB,
+            uint level)
+        {
+            if(string.IsNullOrEmpty(outputFile))
+            {
+                throw new ArgumentNullException(nameof(outputFile));
+            }
+            if(circularBufferSizeInMB == 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(circularBufferSizeInMB));
+            }
+            // TODO: Validate level.
+            // TODO: Consider using an enum.
+            m_outputFile = outputFile;
+            m_circularBufferSizeInMB = circularBufferSizeInMB;
+            m_level = level;
+            m_providers = new List<EventPipeProviderConfiguration>();
+        }
+
+        internal string OutputFile
+        {
+            get { return m_outputFile; }
+        }
+
+        internal uint CircularBufferSizeInMB
+        {
+            get { return m_circularBufferSizeInMB; }
+        }
+
+        internal uint Level
+        {
+            get { return m_level; }
+        }
+
+        internal EventPipeProviderConfiguration[] Providers
+        {
+            get { return m_providers.ToArray(); }
+        }
+
+        internal void EnableProvider(string providerName, UInt64 keywords)
+        {
+            m_providers.Add(new EventPipeProviderConfiguration(
+                providerName,
+                keywords));
+        }
+    }
+
+    internal static class EventPipe
+    {
+        internal static void Enable(EventPipeConfiguration configuration)
+        {
```

Comment:
I notice that EventPipe is a singleton (it has no 'handle' that represents opened pipe).   I think that this will become a needless stumbling block if we ever need more than client.   Lifetime issues also tend to be handled sloppily when you have static singletons (the obligation to close the instance on end of use is not as clear).  Thus I believe that EventPipe be a true instance class, and that the Pinvokes take a handle.  

Note that it is OK in the short term to fail if you try to enable more than one EventPipe simultaneously (although my guess is that you won't need to do that, supporting multiple instances will all naturally 'fall out'.